src/Workspaces/Core/Portable/Workspace/Solution/ProjectOrDocumentId.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal readonly struct ProjectOrDocumentId : IEquatable<ProjectOrDocumentId>
+    {
+        private readonly object? _projectOrDocumentId;
```

Comment:
>_projectOrDocumentId [](start = 33, length = 20)

Doesn't seem like we need to represent `null` value here. I'd prefer declaring `default(ProjectOrDocumentId)` as invalid and making this non-null.

Reply:
The language cannot represent the concept of non-defaultable value types. Until that time, I prefer to harden types unless against this unless it is not possible.

Reply:
I don't think we should make all reference-typed fields of all structs nullable.

Reply:
We should express intent, not strictly the reality as one can always circumvent compiler nullability checks and assigned null to non-nullable variable. The intent here is that `ProjectOrDocumentId` wraps a DocumentId or ProjectId, but never null.

Reply:
The intent you describe is already accurately reflected in the public API for the type.

Reply:
@jaredpar @agocke for guidelines. My understanding is that it is not desirable to mark reference-typed struct fields as nullable if the only way how they can be null is creating a default value of the struct (by calling parameterless ctor or `default(T)`).